### ISSUE TITLE
App Information cut off at bottom on small screen (2)

### ISSUE THREAD

**Issue by vaubaehn** [corona-warn-app/cwa-app-android#3059]
## Avoid duplicates

* [X] Bug is not mentioned in the [FAQ](https://www.coronawarn.app/en/faq/)
* [X] Bug is specific for Android only, for general issues / questions that apply to iOS and Android please raise them in the [documentation repository](https://github.com/corona-warn-app/cwa-documentation)
* [ ] Bug is not already reported in another issue - Bug was already reported in #2302, but can not be reproduced with that hardware configuration there, hence was not reopened.

## Technical details

- Device name: LG-G4 (LGH-815)
- Android version: 6.0
- App version: 2.1.2
- ENS version: 18211213000

## Describe the bug

The bottom of the App Information is cut off on a smaller device and cannot be viewed. The app version information/ENS version information is not displayed. The screen cannot be scrolled.

## Steps to reproduce the issue

1. Tap three-dot overflow menu > App Information
2. Try to scroll the view
![Screenshot_2021-05-03-21-15-07](https://user-images.githubusercontent.com/68376758/116921889-e8544800-ac54-11eb-93ad-4db8eccb50ac.png)

## Expected behaviour

It should be possible to view the version number by scrolling down. The version number is only partially displayed on this device.

## Possible Fix

Possibly similar to #2267? Although there were some other changes in #2501 and #2884 .

## Additional context

After installing the update to CWA 2.1.2, I opened CWA by chosing it from 'app selector' (square right down in device's menu). It started directly from the main screen (MainActivity) without displaying the 'New Features' screen... But I don't think that this messed up the scrolling issue in information fragment. The 'New Features' were displayed after closing CWA and reopening from app drawer.

---
Internal Tracking ID: [EXPOSUREAPP-7040](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-7040)

**Comment by MikeMcC399** [comment]
@vaubaehn 

I will check if I can reproduce with the Nexus 5 emulator on Android 6. Maybe we can find out why it works on Nexus 5, Android 11, but not on your environment.

**Comment by vaubaehn** [comment]
@MikeMcC399 Thank you anyway. So it seems to be API level related (although I guess @fynngodau should have found out already, when preparing for Android 5 compatibility), or there is a different problem...

**Comment by fynngodau** [comment]
I don't think this is, in general, API related.

@vaubaehn Could you confirm whether the other screens mentioned in #2884 are showing padding as in the "Before" screenshots or overscroll as in "After" screenshots for you (or whether there are any patterns you can spot for misbehavior)?

**Comment by vaubaehn** [comment]
@fynngodau thanks for having a look at it. I had a fast check: "Edit Places" screen for example shows the padding 'after', with the blue ripple effect just like in your screenshot. Whereas "App information" shows padding like 'before' and with a small grey circle ripple effect on 'imprint' that is cut off/not overlapping the version number.
Unfortunately it's not possible to take a screenshot of these effects, as my phone is too slow...

**Comment by fynngodau** [comment]
It looks like this is an issue with specific screen sizes. I could reproduce it in the emulator:

![Screenshot_1620110488](https://user-images.githubusercontent.com/16943720/116968350-98a56900-acb4-11eb-8b83-8470c8eb9954.png) | ![Screenshot_1620110556](https://user-images.githubusercontent.com/16943720/116968404-af4bc000-acb4-11eb-8df2-8d7f9578d0df.png)
---|---

**Comment by MikeMcC399** [comment]
@fynngodau 
> It looks like this is an issue with specific screen sizes. I could reproduce it in the emulator:

Which device are you emulating?

**Comment by fynngodau** [comment]
@MikeMcC399 
In this zip file, you can find a device hardware profile for the emulator: [device.zip](https://github.com/corona-warn-app/cwa-app-android/files/6419403/device.zip)

Set the screen size to large.

It does not happen on the Android 9 emulator.

**Comment by fynngodau** [comment]
I'm assuming this to be a platform bug. It appears to happen due to the `fillViewport` attribute; removing it works around the bug:

```
diff --git a/Corona-Warn-App/src/main/res/layout/fragment_information.xml b/Corona-Warn-App/src/main/res/layout/fragment_information.xml
index 645ef7560..b805a54eb 100644
--- a/Corona-Warn-App/src/main/res/layout/fragment_information.xml
+++ b/Corona-Warn-App/src/main/res/layout/fragment_information.xml
@@ -25,7 +25,6 @@
             android:id="@+id/scrollview"
             android:layout_width="match_parent"
             android:layout_height="0dp"
-            android:fillViewport="true"
             android:paddingTop="@dimen/spacing_normal"
             android:paddingBottom="@dimen/spacing_normal"
             android:clipToPadding="false"
```

It would have to be considered which other screens also need to be changed this way.

**Comment by vaubaehn** [comment]
> I'm assuming this to be a platform bug. It appears to happen due to the `fillViewport` attribute; 

Sometimes I don't like my phone to be that 'special'...

> It would have to be considered which other screens also need to be changed this way.

Up to now I couldn't find any other screen to be affected. But it needed to be kept in mind for future changes 😢 

There are still some of these devices out there with CWA installed (according to device specific reviews found in Google Play Store).

Thanks for all your efforts... ❤️ 

By the way, with 2.1.2 all vector drawables render correctly on this device - thanks a lot for this, too!

**Comment by fynngodau** [comment]
The bug is not specific to your device @vaubaehn, it is a bug in the Android platform that was fixed somewhere between Android 6 and Android 9. It happens only if there is *just* not enough space to render the layout without scrolling (full content > screen size but full content < screen size + padding), likely for all other scroll views with padding too. Probably an oversight in the way that it calculates where the padding should go

**Comment by vaubaehn** [comment]
> The bug is not specific to your device @vaubaehn, it is a bug in the Android platform

At least ;)

> It happens only if there is _just_ not enough space to render the layout without scrolling (full content > screen size but full content < screen size + padding)

Did I get you right, that it could easily be worked around, if we added some more content at the bottom of the screen?
In that case, I had an idea, actually... https://github.com/corona-warn-app/cwa-wishlist/issues/497

**Comment by fynngodau** [comment]
> Did I get you right, that it could easily be worked around, if we added some more content at the bottom of the screen?

For your specific screen size + font size yes, but then the problem would happen for a different configuration

**Comment by vaubaehn** [comment]
> For your specific screen size + font size yes, but then the problem would happen for a different configuration

I was just thinking about that, too... would be a bit egoistic...

**Comment by dsarkar** [comment]
@vaubaehn @MikeMcC399 @fynngodau Thanks. Internal Tracking ID: [EXPOSUREAPP-7040](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-7040)

---
Corona-Warn-App Open Source Team

**Comment by MikeMcC399** [comment]
> @fynngodau 
> In this zip file, you can find a device hardware profile (`device.xml`) for the emulator: [device.zip](https://github.com/corona-warn-app/cwa-app-android/files/6419403/device.zip)
> 
> Set the font size to Huge.

Thanks! I was able to reproduce with those settings on Android 6.

**Comment by vaubaehn** [comment]
Hi @dsarkar , you set the "Fix 2.3" label for this issue - could you point to the PR that is supposed to fix the issue? I couldn't find any fitting. Thank you :)

**Comment by dsarkar** [comment]
Hi @vaubaehn, no PR so far available. Label means that issue is (was!) in the pipeline to be fixed by then.

**Comment by dsarkar** [comment]
Forwarded explicitly @fynngodau's analysis to the internal ticket. Thanks !

**Comment by vaubaehn** [comment]
@dsarkar 
Hi Dipankar, thanks for your clarification! Have a nice day!

**Comment by SamuraiKek** [comment]
Hi @fynngodau, I've downloaded your device hardware profile and tried to reproduce this issue on an emulator. I've set the Android version to 6.0 and put the font size to Huge as you've described but everything looked fine, the app version wasn't being cut off, though the ENF version wasn't being displayed, but that's because the hardware profile doesn't support play services. I've forced the ENF version text to show up and checked it once again but it still displayed correctly. I've also created another emulator with 720x720 screen resolution and Android 6.0 in the hopes that the lower dpi would make any difference, but it doesn't. Not sure what else to try, am I missing something? At this point I'm willing to take your word for it and just remove the `fillViewPort` attribute. It's not affecting the view in any way and if there's a chance it'll fix this issue on devices that have this problem, I'll take it.

**Comment by MikeMcC399** [comment]
@SamuraiKek 

The exact conditions for the problem to be visible are quite sensitive. I could reproduce using  @fynngodau's [device.zip](https://github.com/corona-warn-app/cwa-app-android/files/6419403/device.zip) with the released 2.1.2 software, however I could not see the problem when running 2.2.0-RC5 DeviceForTestersDebug (presumably because the menu is longer with the extra "Error Reports" entry). It's good hear that you be opening a PR anyway.

**Comment by fynngodau** [comment]
@SamuraiKek Good that you could reproduce it too. You might also want to consider whether there are any other screens where the `fillViewport` attribute could cause this bug – I noticed that there are a lot of occurrences of this attribute throughout the app, but didn't go through them one-by-one so far.

**Comment by SamuraiKek** [comment]
> @SamuraiKek Good that you could reproduce it too. You might also want to consider whether there are any other screens where the `fillViewport` attribute could cause this bug – I noticed that there are a lot of occurrences of this attribute throughout the app, but didn't go through them one-by-one so far.

I'll have a look.

**Comment by MikeMcC399** [comment]
@vaubaehn 
You may be able to temporarily work around the issue by decreasing or increasing the font size on your device until version 2.3 is available.

The fix https://github.com/corona-warn-app/cwa-app-android/pull/3098 has now been merged into [release/2.3.x](https://github.com/corona-warn-app/cwa-app-android/tree/release/2.3.x). I will see if I can verify it. Unfortunately it is dependent on the screen size, so I will have to experiment a little.

**Comment by MikeMcC399** [comment]
I was able to verify using the device profile from https://github.com/corona-warn-app/cwa-app-android/issues/3059#issuecomment-831722437 with the vertical height extended from 1420px to 1520px on Android 6 (API 23) and with font size set to Huge.

Using the current [release/2.2.x](https://github.com/corona-warn-app/cwa-app-android/tree/release/2.2.x) DeviceForTestersDebug build the App Information screen shows "Imprint" as the bottom-most entry and it will not scroll. The same test with the current [release/2.3.x](https://github.com/corona-warn-app/cwa-app-android/tree/release/2.3.x), which contains the fix from #3098, allows scrolling down, and the version number is completely displayed.

Kudos to @fynngodau for his debugging skills and to @SamuraiKek for fixing it! 🚀 

So now we just have to wait for version 2.3 to be released.

**Comment by dsarkar** [comment]
> Unfortunatelly the issue reoccurs with 2.3.2 and with 2.4.0 RC4 too. But it depends on the font size. If it's bigger, the screen becomes scrollable. I tried to understand what's the difference between the old version and the new and it seems that the "black area" with the Android tools becomes higher on the App Info-List (please toggle between the two 2.4 screenshots - green arrow/red arrow)
> The mobile is the Google Pixel.

**Comment by fynngodau** [comment]
Hi @dsarkar, which Android version is this report about? What can be seen in the screenshots that are mentioned in it (i.e. what green / red arrow)?

**Comment by MikeMcC399** [comment]
@dsarkar 
Please share the screen shots. Possibly also open a new issue if this is not exactly the same problem as the original one.

**Comment by dsarkar** [comment]
@fynngodau @MikeMcC399  
- Google Pixel (Android 10)
-  change the buttons to gestures from the system settings -> issue is not there anymore

**Comment by dsarkar** [comment]
![image](https://user-images.githubusercontent.com/1409741/120789076-64ca9700-c531-11eb-9f3e-bfc0bafa73a8.png)

**Comment by fynngodau** [comment]
Interesting, it looks like the condition "screen size + padding > full content > screen size" from above is fulfilled here again. I thought we worked this out to be a platform bug on older Android versions, huh.

**Comment by MikeMcC399** [comment]
@dsarkar 
Which Pixel device variant is this?

**Comment by MikeMcC399** [comment]
Actually this must be an internal build, not one from the community, because it shows an RC version together with an ENF version. If you are only from the community, you cannot build a version like this, because you can't sign a release version. A debug version cannot access the Google API, so you never see the line "ENF version". That means we will not be able to reproduce the exact issue on the same device in the community. It might still be possible to reproduce it by experimenting with different screen sizes and font sizes though.

I guess that @fynngodau has the right feeling for the problem anyway!

**Comment by dsarkar** [comment]
Variant just pixel

**Comment by MikeMcC399** [comment]
@dsarkar 
If the user looks at Settings > About Phone 
there is normally a Model number shown on Android 10.
(Or in German
Einstellungen > Info zu  Telefon / Telefoninfo
dann Modellnummer.)

**Comment by MikeMcC399** [comment]
I tried to reproduce this issue in a Pixel emulator, but was unable to do so. As we know from previously, the issue can be very sensitive to exact screen layouts. The API 29 (Android 10) emulators don't have Bluetooth. The debug BuildTypes don't have access to the ENS API. That means I can't exactly recreate a Pixel, Android 10 environment for 2.3.x or 2.4.x with access to ENS API. Hopefully the Open Source Team can recreate the issue internally and look at fixing it.

**Comment by MikeMcC399** [comment]
I got lucky 🙂 and was able to reproduce the issue on a Pixel emulator, Android 10, with font size set to large. I have documented this in #3369 with screen shots.

I suspect that the buttons from the Start Screen are adversely affecting the layout of the App Information view.

**Comment by MikeMcC399** [comment]
The problem, as reported by @vaubaehn, was on Android 6.0 with app version 2.1.2 on an LG-G4 (LGH-815) device.

Here are some up-to-date test results that may help find and solve the issue on Android 6.

Using the emulator hardware profile (`device.xml`) from @fynngodau in the post https://github.com/corona-warn-app/cwa-app-android/issues/3059#issuecomment-831722437
[device.zip](https://github.com/corona-warn-app/cwa-app-android/files/6419403/device.zip).

- With build 2.1.2-RC2-deviceForTesters-debug on Android 6.0 (API 23)
Font size: Large: the App Information is truncated with display including Imprint and nothing further. The version number is not displayed. The view does not scroll.
(Font size: Small or Normal: shows the whole of App Information without scrolling. Font size: Huge: App Information is scrollable.)

- With build 2.4.0-RC5-deviceForTesters-debug on Android 6.0 (API 23)
Font size: Normal: the App Information is truncated with partial display of the app version number. The view does not scroll normally. It can however be made to scroll by selecting and dragging the area around the truncated version information.
(Font size: Small: shows the whole of App Information without scrolling. Font size: Large or Huge: App Information is scrollable by tapping and dragging anywhere inside the App Information view.)

**Comment by MikeMcC399** [comment]
I've suggested in https://github.com/corona-warn-app/cwa-app-android/issues/3369 to add some additional space at the bottom of the scrolled App Information view as a workaround. That seems to work on a few different configurations I tried. 

I'd be happy to submit this as a PR but I'd like some dev feedback first before I write up a PR. Also, since Internal Tracking ID: [EXPOSUREAPP-7040](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-7040) has been used for PR https://github.com/corona-warn-app/cwa-app-android/pull/3098 in [v2.3.0-RC0-0](https://github.com/corona-warn-app/cwa-app-android/releases/tag/v2.3.0-RC0-0) it would probably be good to get a new Internal Tracking ID for any additional fix.

**Comment by MikeMcC399** [comment]
I've submitted the workaround, which I mentioned in the previous post, as PR #3416.

**Comment by MikeMcC399** [comment]
@vaubaehn 

Could you let us know what your experience is with the latest release 2.3.2 on your LG-G4 (LGH-815) / Android version: 6.0 phone?

Can you now scroll the "App Information" screen easily?
If the scrolling doesn't work properly, what happens if you tap, hold and drag the truncated version information upwards?
What is your font size set to?

If your device scrolls correctly, I would suggest closing this issue and following https://github.com/corona-warn-app/cwa-app-android/issues/3369 only. On the other hand, if you are still having issues with scrolling, then I would suggest leaving your issue open as well.

The workaround https://github.com/corona-warn-app/cwa-app-android/pull/3416 I submitted has been merged into the [release/2.4.x](https://github.com/corona-warn-app/cwa-app-android/tree/release/2.4.x) branch, so hopefully your issue should be solved at the latest with the release of 2.4.x.

**Comment by vaubaehn** [comment]
@MikeMcC399 
> Can you now scroll the "App Information" screen easily?

Yes, it's working.

> What is your font size set to?

I was using font size large with CWA 2.2.1, which let me even more feel like a senior than I actually am currently - but it was working well as a workaround for being able to scroll.

Now, with CWA 2.3.2, I set font size back to medium (normal, standard), and I'm able to scroll without any issue.

> If your device scrolls correctly, I would suggest closing this issue and following #3369 only.

So will I do.

Thanks for all your engagement!
